drivers/sensor/adxl372/adxl372.c

```diff
@@ -0,0 +1,1081 @@
+{
+	u8_t buf[1024];
+	u16_t i;
+	int ret;
+
+	if (cnt > 512 || cnt == 0) {
+		return -EINVAL;
+	}
+	/*
+	 * The FIFO can hold up to 512 samples.
+	 * Each sample is 2 bytes, that's why we read (cnt * 2) bytes
+	 */
+	ret = adxl372_reg_read_multiple(dev, ADXL372_FIFO_DATA, buf, cnt * 2);
+
+	for (i = 0; i < cnt * 2; i += 6) {
+		samples->x = ((buf[i] << 8) | (buf[i+1] & 0xF0));
+		samples->y = ((buf[i+2] << 8) | (buf[i+3] & 0xF0));
+		samples->z = ((buf[i+4] << 8) | (buf[i+5] & 0xF0));
+
+		samples++;
+	}
+
+	return ret;
+}
+
+/**
+ * Configure the operating parameters for the FIFO.
+ * @param dev - The device structure.
+ * @param mode - FIFO Mode. Specifies FIFO operating mode.
+ *		Accepted values: ADXL372_FIFO_BYPASSED
+ *				 ADXL372_FIFO_STREAMED
+ *				 ADXL372_FIFO_TRIGGERED
+ *				 ADXL372_FIFO_OLD_SAVED
+ * @param format - FIFO Format. Specifies the data is stored in the FIFO buffer.
+ *		Accepted values: ADXL372_XYZ_FIFO
+ *				 ADXL372_X_FIFO
+ *				 ADXL372_Y_FIFO
+ *				 ADXL372_XY_FIFO
+ *				 ADXL372_Z_FIFO
+ *				 ADXL372_XZ_FIFO
+ *				 ADXL372_YZ_FIFO
+ *				 ADXL372_XYZ_PEAK_FIFO
+ * @param fifo_samples - FIFO Samples. Watermark number of FIFO samples that
+ *			triggers a FIFO_FULL condition when reached.
+ *			Values range from 0 to 512.
+
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int adxl372_configure_fifo(struct device *dev,
+			       enum adxl372_fifo_mode mode,
+			       enum adxl372_fifo_format format,
+			       u16_t fifo_samples)
+{
+	struct adxl372_data *data = dev->driver_data;
+	u8_t fifo_config;
+	int ret;
+
+	if (fifo_samples > 512) {
+		return -EINVAL;
+	}
+
+	/*
+	 * All FIFO modes must be configured while in standby mode.
+	 */
+	ret = adxl372_set_op_mode(dev, ADXL372_STANDBY);
+	if (ret) {
+		return ret;
+	}
+
+	fifo_config = (ADXL372_FIFO_CTL_FORMAT_MODE(format) |
+		       ADXL372_FIFO_CTL_MODE_MODE(mode) |
+		       ADXL372_FIFO_CTL_SAMPLES_MODE(fifo_samples));
+
+	ret = adxl372_reg_write(dev, ADXL372_FIFO_CTL, fifo_config);
+	if (ret) {
+		return ret;
+	}
+	ret = adxl372_reg_write(dev, ADXL372_FIFO_SAMPLES, fifo_samples & 0xFF);
+	if (ret) {
+		return ret;
+	}
+
+	data->fifo_config.fifo_format = format;
+	data->fifo_config.fifo_mode = mode;
+	data->fifo_config.fifo_samples = fifo_samples;
+
+	return 0;
+}
+
+/**
+ * Retrieve data stored in FIFO. Can be used in polling mode,
+ * but works best when interrupts are used
+ * @param dev - The device structure.
+ * @param fifo_data - pointer to an array of type adxl372_xyz_accel_data
+ *		      where (x, y, z) values will be stored. Array max size
+ *		      should be 170, as the FIFO can hold up to 512 samples.
+ * @param fifo_entries - pointer which will store the number of valid data
+ *			 samples present in the FIFO buffer
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int adxl372_service_fifo_ev(struct device *dev,
+			    struct adxl372_xyz_accel_data *fifo_data,
+				u16_t *fifo_entries)
```

Comment:
Not yet - I have a second patch-set or RFC waiting but wanted to get the driver submitted first.

The FIFO buffer is an important feature in ultralow power applications in two ways:

The FIFO enables system level power savings by enabling the host processor to sleep for extended periods while the accelerometer autonomously collects data. Alternatively, using the FIFO to collect data can unburden the host while it tends to other tasks.

The FIFO can be used in a triggered mode to record all data leading up to an activity detection event, thereby providing context for the event. In the case of a system that identifies impact events, for example, the accelerometer can keep the entire system off while it stores acceleration data in its FIFO and looks for an activity event. When the impact event occurs, data collected prior to the event is frozen in the FIFO. The accelerometer can now trigger/wake the rest of the system and transfer this data to the host processor, thereby providing context for the impact event.

However to properly utilize the FIFO we need to extend the sensor_driver_api with a new channel_get method where the user specifies how many samples it can accept and then the method would return how many samples it actually returns.

Reply:
I agree FIFO support is a valuable feature (see #1387), but please wait to add this function (and any related helper functions) until we have some way to use it. An RFC for the sensor_driver_api would be welcome.